File: helper-code/tack-and-strat-java/src/main/java/com/tackstrat/model/WorldGenerator.java
package com.tackstrat.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

/** Continents, mountain ridges, desert patches, and spaced starts. */
public final class WorldGenerator {

    private static final int MAP_RADIUS = 16;
    private static final int CONTINENT_TARGET_TILES = 90;

    private WorldGenerator() {}

    public static GeneratedWorld generate(List<Player> players, long seed) {
        var rnd = new Random(seed);
        var builder = new GameMap.Builder(MAP_RADIUS);

        for (var c : new ArrayList<>(builder.mutableCells().keySet())) {
            builder.set(c, Terrain.WATER);
        }

        // 1. Continents (one per player + 1 wild)
        int continents = Math.max(2, players.size() + 1);
        List<HexCoord> seeds = pickContinentSeeds(builder, rnd, continents);
        Map<HexCoord, Integer> tileToContinent = new HashMap<>();
        for (int idx = 0; idx < seeds.size(); idx++) {
            growContinent(builder, seeds.get(idx), idx, tileToContinent, rnd);
        }

        // 2. Biomes for land tiles (coast/inland bias)
        for (var c : new ArrayList<>(builder.mutableCells().keySet())) {
            if (builder.mutableCells().get(c) != Terrain.WATER) {
                builder.set(c, pickBiome(c, builder, rnd));
            }
        }

        // 3. Mountain ridges (run a chain of mountains across some inland tiles)
        carveMountainRidges(builder, tileToContinent, rnd);

        // 4. Desert patches (small clusters in inland / non-forest)
        carveDeserts(builder, tileToContinent, rnd);

        var map = builder.build();
        List<HexCoord> spawns = pickSpawnsByContinent(map, tileToContinent, players.size(), rnd);

        var units = new ArrayList<Unit>();
        int uid = 1;
        for (int i = 0; i < players.size(); i++) {
            var p = players.get(i);
            HexCoord home = spawns.get(i);
            units.add(new Unit(uid++, p.seat(), UnitKind.SETTLER, home));
            // Each player also gets a Scout one tile away if possible
            HexCoord scoutCoord = pickAdjacent(home, map, units);
            if (scoutCoord != null) {
                units.add(new Unit(uid++, p.seat(), UnitKind.SCOUT, scoutCoord));
            }
        }

        Map<HexCoord, Integer> soil = new HashMap<>();
        for (HexCoord c : map.passableLand()) {
            soil.put(c, rnd.nextInt(3)); // 0–2 natural fertility
        }
        // Wildlife is spawned during play (herds or loners, irregular timing) — see GameSession.
        return new GeneratedWorld(map, units, soil, List.of());
    }

    private static List<HexCoord> pickContinentSeeds(GameMap.Builder builder, Random rnd, int n) {
        int innerRadius = Math.max(3, (MAP_RADIUS * 7) / 10);
        var candidates = new ArrayList<HexCoord>();
        for (var c : builder.mutableCells().keySet()) {
            if (Math.abs(c.q()) <= innerRadius
                    && Math.abs(c.r()) <= innerRadius
                    && Math.abs(c.q() + c.r()) <= innerRadius) {
                candidates.add(c);
            }
        }
        Collections.shuffle(candidates, rnd);

        var picked = new ArrayList<HexCoord>();
        int minDist = Math.max(5, MAP_RADIUS - 6);
        for (var c : candidates) {
            boolean ok = true;
            for (var p : picked) {
                if (p.distanceTo(c) < minDist) {
                    ok = false;
                    break;
                }
            }
            if (ok) {
                picked.add(c);
                if (picked.size() == n) {
                    break;
                }
            }
        }
        for (var c : candidates) {
            if (picked.size() == n) {
                break;
            }
            if (!picked.contains(c)) {
                picked.add(c);
            }
        }
        return picked;
    }

    private static void growContinent(
            GameMap.Builder builder,
            HexCoord seed,
            int continentId,
            Map<HexCoord, Integer> tileToContinent,
            Random rnd) {
        var cells = builder.mutableCells();
        if (!cells.containsKey(seed)) {
            return;
        }
        int target = CONTINENT_TARGET_TILES + rnd.nextInt(40) - 20;
        int converted = 0;

        var frontier = new ArrayList<HexCoord>();
        frontier.add(seed);

        while (converted < target && !frontier.isEmpty()) {
            int weight = Math.min(frontier.size(), 12);
            int pickIdx = rnd.nextInt(weight);
            var c = frontier.remove(pickIdx);

            if (cells.get(c) != Terrain.WATER || tileToContinent.containsKey(c)) {
                continue;
            }
            // Keep an outer ocean ring for cleaner coastlines
            if (Math.abs(c.q()) >= MAP_RADIUS
                    || Math.abs(c.r()) >= MAP_RADIUS
                    || Math.abs(c.q() + c.r()) >= MAP_RADIUS) {
                continue;
            }

            cells.put(c, Terrain.GRASS);
            tileToContinent.put(c, continentId);
            converted++;

            for (var n : c.neighbors()) {
                if (cells.containsKey(n)
                        && cells.get(n) == Terrain.WATER
                        && !tileToContinent.containsKey(n)) {
                    frontier.add(n);
                }
            }
            if (rnd.nextInt(4) == 0) {
                Collections.shuffle(frontier, rnd);
            }
        }
    }

    private static Terrain pickBiome(HexCoord c, GameMap.Builder builder, Random rnd) {
        var cells = builder.mutableCells();
        int waterNeighbors = 0;
        int existing = 0;
        for (var n : c.neighbors()) {
            if (!cells.containsKey(n)) {
                waterNeighbors++;
                continue;
            }
            existing++;
            if (cells.get(n) == Terrain.WATER) {
                waterNeighbors++;
            }
        }
        boolean coastal = waterNeighbors >= 2 || existing < 6;

        double roll = rnd.nextDouble();
        if (coastal) {
            if (roll < 0.55) return Terrain.GRASS;
            if (roll < 0.85) return Terrain.PLAINS;
            return Terrain.FOREST;
        }
        if (roll < 0.28) return Terrain.GRASS;
        if (roll < 0.55) return Terrain.PLAINS;
        if (roll < 0.85) return Terrain.FOREST;
        return Terrain.HILL;
    }

    private static void carveMountainRidges(
            GameMap.Builder builder,
            Map<HexCoord, Integer> tileToContinent,
            Random rnd) {
        var cells = builder.mutableCells();
        // Group land tiles by continent
        Map<Integer, List<HexCoord>> byContinent = new HashMap<>();
        for (var e : tileToContinent.entrySet()) {
            byContinent.computeIfAbsent(e.getValue(), k -> new ArrayList<>()).add(e.getKey());
        }
        for (var continent : byContinent.values()) {
            if (continent.size() < 25) {
                continue;
            }
            int ridgeLength = 4 + rnd.nextInt(5);
            // Start ridge at an inland tile
            HexCoord start = inlandTile(continent, cells, rnd);
            if (start == null) {
                continue;
            }
            HexCoord cursor = start;
            // Pick a primary direction (axial step)
            HexCoord[] dirs = {
                    new HexCoord(1, 0), new HexCoord(1, -1), new HexCoord(0, -1),
                    new HexCoord(-1, 0), new HexCoord(-1, 1), new HexCoord(0, 1),
            };
            HexCoord dir = dirs[rnd.nextInt(6)];
            for (int i = 0; i < ridgeLength; i++) {
                if (cells.containsKey(cursor)
                        && cells.get(cursor) != Terrain.WATER
                        && cells.get(cursor) != Terrain.MOUNTAIN) {
                    cells.put(cursor, Terrain.MOUNTAIN);
                }
                // Occasional jog
                HexCoord step = (rnd.nextInt(4) == 0)
                        ? dirs[rnd.nextInt(6)]
                        : dir;
                cursor = cursor.add(step);
            }
        }
    }

    private static HexCoord inlandTile(List<HexCoord> tiles, Map<HexCoord, Terrain> cells, Random rnd) {
        var copy = new ArrayList<>(tiles);
        Collections.shuffle(copy, rnd);
        for (var c : copy) {
            int waterNeighbors = 0;
            for (var n : c.neighbors()) {
                if (!cells.containsKey(n) || cells.get(n) == Terrain.WATER) {
                    waterNeighbors++;
                }
            }
            if (waterNeighbors == 0) {
                return c;
            }
        }
        return copy.isEmpty() ? null : copy.get(0);
    }

    private static void carveDeserts(
            GameMap.Builder builder,
            Map<HexCoord, Integer> tileToContinent,
            Random rnd) {
        var cells = builder.mutableCells();
        Map<Integer, List<HexCoord>> byContinent = new HashMap<>();
        for (var e : tileToContinent.entrySet()) {
            byContinent.computeIfAbsent(e.getValue(), k -> new ArrayList<>()).add(e.getKey());
        }
        for (var continent : byContinent.values()) {
            if (continent.size() < 30) {
                continue;
            }
            // 1-2 desert blobs per continent
            int blobs = 1 + rnd.nextInt(2);
            for (int b = 0; b < blobs; b++) {
                HexCoord seed = inlandTile(continent, cells, rnd);
                if (seed == null) {
                    continue;
                }
                int target = 4 + rnd.nextInt(5);
                var frontier = new ArrayList<HexCoord>();
                frontier.add(seed);
                int converted = 0;
                while (converted < target && !frontier.isEmpty()) {
                    var c = frontier.remove(rnd.nextInt(frontier.size()));
                    if (!cells.containsKey(c)) continue;
                    Terrain t = cells.get(c);
                    if (t == Terrain.WATER || t == Terrain.MOUNTAIN || t == Terrain.DESERT) {
                        continue;
                    }
                    cells.put(c, Terrain.DESERT);
                    converted++;
                    for (var n : c.neighbors()) {
                        if (cells.containsKey(n) && cells.get(n) != Terrain.WATER) {
                            frontier.add(n);
                        }
                    }
                }
            }
        }
    }

    private static List<HexCoord> pickSpawnsByContinent(
            GameMap map,
            Map<HexCoord, Integer> tileToContinent,
            int n,
            Random rnd) {
        Map<Integer, List<HexCoord>> byContinent = new HashMap<>();
        for (var c : map.passableLand()) {
            int cid = tileToContinent.getOrDefault(c, -1);
            byContinent.computeIfAbsent(cid, k -> new ArrayList<>()).add(c);
        }
        var continents = new ArrayList<>(byContinent.values());
        continents.removeIf(List::isEmpty);
        continents.sort(Comparator.<List<HexCoord>>comparingInt(List::size).reversed());

        var out = new ArrayList<HexCoord>();
        Set<Integer> used = new HashSet<>();
        for (var continent : continents) {
            if (out.size() == n) break;
            HexCoord pick = pickInteriorTile(continent, map);
            if (pick != null) {
                out.add(pick);
                used.add(System.identityHashCode(continent));
            }
        }
        // Fallback
        if (out.size() < n) {
            var fallback = new ArrayList<>(map.passableLand());
            Collections.shuffle(fallback, rnd);
            for (var c : fallback) {
                if (!out.contains(c)) {
                    out.add(c);
                    if (out.size() == n) break;
                }
            }
        }
        if (out.size() < n) {
            throw new IllegalStateException("Not enough land for " + n + " players");
        }
        return out;
    }

    private static HexCoord pickInteriorTile(List<HexCoord> tiles, GameMap map) {
        HexCoord best = null;
        int bestLand = -1;
        for (var c : tiles) {
            if (!map.terrainAt(c).canFoundCityOn()) {
                continue;
            }
            int land = 0;
            for (var nb : c.neighbors()) {
                if (map.contains(nb) && map.terrainAt(nb).passable()) {
                    land++;
                }
            }
            if (land > bestLand) {
                bestLand = land;
                best = c;
            }
            if (bestLand == 6) break;
        }
        if (best == null && !tiles.isEmpty()) {
            best = tiles.get(0);
        }
        return best;
    }

    private static HexCoord pickAdjacent(HexCoord c, GameMap map, List<Unit> existing) {
        for (var n : c.neighbors()) {
            if (!map.contains(n) || !map.terrainAt(n).passable()) continue;
            boolean occupied = false;
            for (var u : existing) {
                if (u.coord().equals(n)) {
                    occupied = true;
                    break;
                }
            }
            if (!occupied) {
                return n;
            }
        }
        return null;
    }
}

File: src/engine/world_generator.cpp
#include "tack/strat/world_generator.hpp"

#include <algorithm>
#include <optional>
#include <stdexcept>
#include <unordered_map>
#include <vector>

#include "tack/strat/java_random.hpp"
#include "tack/strat/map_size.hpp"

namespace tack::strat {

namespace {

int constexpr kContinentTargetTiles = 90;

std::vector<HexCoord> pick_continent_seeds(GameMap::Builder& builder, JavaRandom& rnd, int n,
                                           int cols, int rows) {
  int const span = std::max(8, std::min(cols, rows));
  int const inner_d = std::max(3, (span * 7) / 10);
  std::vector<HexCoord> candidates;
  for (auto const& e : builder.mutable_cells()) {
    HexCoord const c = e.first;
    int const d = std::max({std::abs(c.q), std::abs(c.r), std::abs(c.q + c.r)});
    if (d <= inner_d) {
      candidates.push_back(c);
    }
  }
  java_collections_shuffle(candidates, rnd);

  std::vector<HexCoord> picked;
  int const min_dist = std::max(5, span / 3);
  for (HexCoord const c : candidates) {
    bool ok = true;
    for (HexCoord const p : picked) {
      if (p.distance_to(c) < min_dist) {
        ok = false;
        break;
      }
    }
    if (ok) {
      picked.push_back(c);
      if (static_cast<int>(picked.size()) == n) {
        break;
      }
    }
  }
  for (HexCoord const c : candidates) {
    if (static_cast<int>(picked.size()) == n) {
      break;
    }
    if (std::find(picked.begin(), picked.end(), c) == picked.end()) {
      picked.push_back(c);
    }
  }
  return picked;
}

void grow_continent(GameMap::Builder& builder, HexCoord seed, int continent_id,
                    std::unordered_map<HexCoord, int>& tile_to_continent, JavaRandom& rnd) {
  auto& cells = builder.mutable_cells();
  int target = kContinentTargetTiles + rnd.next_int_bounded(40) - 20;
  int converted = 0;
  std::vector<HexCoord> frontier;
  frontier.push_back(seed);

  while (converted < target && !frontier.empty()) {
    int weight = std::min(static_cast<int>(frontier.size()), 12);
    int pick_idx = rnd.next_int_bounded(weight);
    HexCoord c = frontier[static_cast<std::size_t>(pick_idx)];
    frontier.erase(frontier.begin() + pick_idx);

    auto it = cells.find(c);
    if (it == cells.end() || it->second != Terrain::WATER || tile_to_continent.count(c)) {
      continue;
    }

    it->second = Terrain::GRASS;
    tile_to_continent[c] = continent_id;
    converted++;

    for (HexCoord n : c.neighbors()) {
      auto j = cells.find(n);
      if (j != cells.end() && j->second == Terrain::WATER && !tile_to_continent.count(n)) {
        frontier.push_back(n);
      }
    }
    if (rnd.next_int_bounded(4) == 0) {
      java_collections_shuffle(frontier, rnd);
    }
  }
}

Terrain pick_biome(HexCoord c, GameMap::Builder& builder, JavaRandom& rnd) {
  auto& cells = builder.mutable_cells();
  int water_neighbors = 0;
  int existing = 0;
  for (HexCoord n : c.neighbors()) {
    auto j = cells.find(n);
    if (j == cells.end()) {
      water_neighbors++;
      continue;
    }
    existing++;
    if (j->second == Terrain::WATER) {
      water_neighbors++;
    }
  }
  bool coastal = water_neighbors >= 2 || existing < 6;

  double roll = rnd.next_double();
  if (coastal) {
    if (roll < 0.55) return Terrain::GRASS;
    if (roll < 0.85) return Terrain::PLAINS;
    return Terrain::FOREST;
  }
  if (roll < 0.28) return Terrain::GRASS;
  if (roll < 0.55) return Terrain::PLAINS;
  if (roll < 0.85) return Terrain::FOREST;
  return Terrain::HILL;
}

std::optional<HexCoord> inland_tile(std::vector<HexCoord> const& tiles,
                                    std::unordered_map<HexCoord, Terrain> const& cells,
                                    JavaRandom& rnd) {
  auto copy = tiles;
  java_collections_shuffle(copy, rnd);
  for (HexCoord const c : copy) {
    int water_neighbors = 0;
    for (HexCoord n : c.neighbors()) {
      auto j = cells.find(n);
      if (j == cells.end() || j->second == Terrain::WATER) {
        water_neighbors++;
      }
    }
    if (water_neighbors == 0) {
      return c;
    }
  }
  if (!copy.empty()) {
    return copy.front();
  }
  return std::nullopt;
}

void carve_mountain_ridges(GameMap::Builder& builder,
                           std::unordered_map<HexCoord, int> const& tile_to_continent,
                           JavaRandom& rnd) {
  auto& cells = builder.mutable_cells();
  std::unordered_map<int, std::vector<HexCoord>> by_continent;
  for (auto const& e : tile_to_continent) {
    by_continent[e.second].push_back(e.first);
  }
  for (auto& entry : by_continent) {
    auto& continent = entry.second;
    if (continent.size() < 25) {
      continue;
    }
    int ridge_length = 4 + rnd.next_int_bounded(5);
    auto start_opt = inland_tile(continent, cells, rnd);
    if (!start_opt) {
      continue;
    }
    HexCoord cursor = *start_opt;
    HexCoord const dirs[] = {HexCoord{1, 0}, HexCoord{1, -1}, HexCoord{0, -1},
                             HexCoord{-1, 0}, HexCoord{-1, 1}, HexCoord{0, 1}};
    HexCoord dir = dirs[rnd.next_int_bounded(6)];
    for (int i = 0; i < ridge_length; ++i) {
      auto it = cells.find(cursor);
      if (it != cells.end() && it->second != Terrain::WATER && it->second != Terrain::MOUNTAIN) {
        it->second = Terrain::MOUNTAIN;
      }
      HexCoord step = (rnd.next_int_bounded(4) == 0) ? dirs[rnd.next_int_bounded(6)] : dir;
      cursor = cursor.add(step);
    }
  }
}

void carve_deserts(GameMap::Builder& builder,
                   std::unordered_map<HexCoord, int> const& tile_to_continent, JavaRandom& rnd) {
  auto& cells = builder.mutable_cells();
  std::unordered_map<int, std::vector<HexCoord>> by_continent;
  for (auto const& e : tile_to_continent) {
    by_continent[e.second].push_back(e.first);
  }
  for (auto& entry : by_continent) {
    auto& continent = entry.second;
    if (continent.size() < 30) {
      continue;
    }
    int blobs = 1 + rnd.next_int_bounded(2);
    for (int b = 0; b < blobs; ++b) {
      auto seed_opt = inland_tile(continent, cells, rnd);
      if (!seed_opt) {
        continue;
      }
      HexCoord seed = *seed_opt;
      int target = 4 + rnd.next_int_bounded(5);
      std::vector<HexCoord> frontier;
      frontier.push_back(seed);
      int converted = 0;
      while (converted < target && !frontier.empty()) {
        int idx = rnd.next_int_bounded(static_cast<int>(frontier.size()));
        HexCoord c = frontier[static_cast<std::size_t>(idx)];
        frontier.erase(frontier.begin() + idx);
        auto it = cells.find(c);
        if (it == cells.end()) continue;
        Terrain t = it->second;
        if (t == Terrain::WATER || t == Terrain::MOUNTAIN || t == Terrain::DESERT) {
          continue;
        }
        it->second = Terrain::DESERT;
        converted++;
        for (HexCoord n : c.neighbors()) {
          auto j = cells.find(n);
          if (j != cells.end() && j->second != Terrain::WATER) {
            frontier.push_back(n);
          }
        }
      }
    }
  }
}

HexCoord pick_interior_tile(std::vector<HexCoord> const& tiles, GameMap const& map) {
  HexCoord best{};
  int best_land = -1;
  for (HexCoord const c : tiles) {
    if (!can_found_city_on(map.terrain_at(c))) {
      continue;
    }
    int land = 0;
    for (HexCoord nb : c.neighbors()) {
      if (map.contains(nb) && passable(map.terrain_at(nb))) {
        land++;
      }
    }
    if (land > best_land) {
      best_land = land;
      best = c;
      if (best_land == 6) {
        break;
      }
    }
  }
  if (best_land < 0 && !tiles.empty()) {
    best = tiles.front();
  }
  return best;
}

std::vector<HexCoord> pick_spawns_by_continent(GameMap const& map,
                                               std::unordered_map<HexCoord, int> const& tile_to_continent,
                                               int n, JavaRandom& rnd) {
  std::unordered_map<int, std::vector<HexCoord>> by_continent;
  for (HexCoord const c : map.passable_land()) {
    int cid = -1;
    auto it = tile_to_continent.find(c);
    if (it != tile_to_continent.end()) {
      cid = it->second;
    }
    by_continent[cid].push_back(c);
  }
  std::vector<std::vector<HexCoord>> continents;
  continents.reserve(by_continent.size());
  for (auto& e : by_continent) {
    if (!e.second.empty()) {
      continents.push_back(std::move(e.second));
    }
  }
  std::sort(continents.begin(), continents.end(),
            [](auto const& a, auto const& b) { return a.size() > b.size(); });

  std::vector<HexCoord> out;
  for (auto const& continent : continents) {
    if (static_cast<int>(out.size()) == n) {
      break;
    }
    HexCoord pick = pick_interior_tile(continent, map);
    out.push_back(pick);
  }
  if (static_cast<int>(out.size()) < n) {
    auto fallback = map.passable_land();
    java_collections_shuffle(fallback, rnd);
    for (HexCoord const c : fallback) {
      if (std::find(out.begin(), out.end(), c) == out.end()) {
        out.push_back(c);
        if (static_cast<int>(out.size()) == n) {
          break;
        }
      }
    }
  }
  if (static_cast<int>(out.size()) < n) {
    throw std::runtime_error("Not enough land for player spawns");
  }
  return out;
}

std::optional<HexCoord> pick_adjacent(HexCoord home, GameMap const& map,
                                      std::vector<Unit> const& existing) {
  for (HexCoord n : home.neighbors()) {
    if (!map.contains(n) || !passable(map.terrain_at(n))) {
      continue;
    }
    bool occupied = false;
    for (Unit const& u : existing) {
      if (u.coord() == n) {
        occupied = true;
        break;
      }
    }
    if (!occupied) {
      return n;
    }
  }
  return std::nullopt;
}

}  // namespace

GeneratedWorld WorldGenerator::generate(std::vector<Player> const& players, std::int64_t seed,
                                        MapSizePreset map_size) {
  JavaRandom rnd(seed);
  MapCivHexDims const dim = map_civ_hex_dims(map_size);
  GameMap::Builder builder(CivHexRectangleTag{}, dim.cols, dim.rows);

  std::vector<HexCoord> cell_keys;
  cell_keys.reserve(builder.mutable_cells().size());
  for (auto const& e : builder.mutable_cells()) {
    cell_keys.push_back(e.first);
  }
  std::sort(cell_keys.begin(), cell_keys.end(),
             [](HexCoord a, HexCoord b) { return a.q != b.q ? a.q < b.q : a.r < b.r; });
  for (HexCoord c : cell_keys) {
    builder.set(c, Terrain::WATER);
  }

  int continents = std::max(2, static_cast<int>(players.size()) + 1);
  auto seeds = pick_continent_seeds(builder, rnd, continents, dim.cols, dim.rows);
  std::unordered_map<HexCoord, int> tile_to_continent;
  for (std::size_t idx = 0; idx < seeds.size(); ++idx) {
    grow_continent(builder, seeds[idx], static_cast<int>(idx), tile_to_continent, rnd);
  }

  std::vector<HexCoord> land_cells;
  for (auto const& e : builder.mutable_cells()) {
    if (e.second != Terrain::WATER) {
      land_cells.push_back(e.first);
    }
  }
  std::sort(land_cells.begin(), land_cells.end(),
            [](HexCoord a, HexCoord b) { return a.q != b.q ? a.q < b.q : a.r < b.r; });
  for (HexCoord c : land_cells) {
    builder.set(c, pick_biome(c, builder, rnd));
  }

  carve_mountain_ridges(builder, tile_to_continent, rnd);
  carve_deserts(builder, tile_to_continent, rnd);

  GameMap map = builder.build();
  auto spawns = pick_spawns_by_continent(map, tile_to_continent, static_cast<int>(players.size()), rnd);

  std::vector<Unit> units;
  int uid = 1;
  for (std::size_t i = 0; i < players.size(); ++i) {
    Player const& p = players[i];
    HexCoord home = spawns[i];
    units.emplace_back(uid++, p.seat, UnitKind::SETTLER, home);
    if (auto scout_coord = pick_adjacent(home, map, units)) {
      units.emplace_back(uid++, p.seat, UnitKind::SCOUT, *scout_coord);
    }
  }

  std::unordered_map<HexCoord, int> soil;
  for (HexCoord c : map.passable_land()) {
    soil[c] = rnd.next_int_bounded(3);
  }

  return GeneratedWorld{std::move(map), std::move(units), std::move(soil), {}};
}

}  // namespace tack::strat
